File: Carros Objective-C/Carros Objective-C/Model/Utils/DownloadImageView.m
//
//  DownloadImageView.m
//  Carros Objective-C
//
//  Created by Antonio Lopes on 8/9/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//

#import "DownloadImageView.h"

@implementation DownloadImageView

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self createProgress];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self createProgress];
    }
    return self;
}

- (void) createProgress {
    progress = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    [self addSubview:progress];
}

- (void) layoutSubviews {
    progress.center = [self convertPoint:self.center toView:self.superview];
}

- (void) setUrl:(NSString*)url {
    [self setUrl:url withCache:true];
}

- (void) setUrl:(NSString*)url withCache:(BOOL)cache {
    self.image = nil;
    [progress startAnimating];
    NSOperationQueue * queue = [[NSOperationQueue alloc] init];
    [queue addOperationWithBlock:^{ [self downloadImg:url withCache:cache]; }];
}

- (void) downloadImg:(NSString *)url withCache:(BOOL)cache {
    NSData * data = nil;
    if (!cache) {
        data = [[NSData alloc] initWithContentsOfURL:[NSURL URLWithString:url]];
    } else {
        NSString * path = url;
        path = [path stringByReplacingOccurrencesOfString:@"/" withString:@"_"];
        path = [path stringByReplacingOccurrencesOfString:@"\\" withString:@"_"];
        path = [path stringByReplacingOccurrencesOfString:@":" withString:@"_"];
        path = [NSString stringWithFormat:@"%@/Documents/%@", NSHomeDirectory(), path];
        
        if ( [[NSFileManager defaultManager] fileExistsAtPath:path] ) {
            data = [[NSData alloc] initWithContentsOfFile:path];
        } else {
            data = [[NSData alloc] initWithContentsOfURL:[NSURL URLWithString:url]];
            [data writeToFile:path atomically:true];
        }
    }
    
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{ [self showImg:data]; }];
}

- (void) showImg:(NSData *)data {
    if (data.length == 0) {
        return;
    }
    
    [self setImage:[UIImage imageWithData:data]];
    [progress stopAnimating];
}

@end

File: Carros Swift/Carros Swift/Model/Utils/DownloadImageView.swift
//
//  DownloadImageView.swift
//  Carros Swift
//
//  Created by Antonio Lopes on 8/9/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//

import UIKit

class DownloadImageView: UIImageView {

    // MARK: Local Variables
    var progress : UIActivityIndicatorView!
    let queue = OperationQueue()
    let mainQueue = OperationQueue.main
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        createProgress()
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        createProgress()
    }
    
    func createProgress() {
        progress = UIActivityIndicatorView(activityIndicatorStyle: UIActivityIndicatorViewStyle.gray)
        addSubview(progress)
    }

    override func layoutSubviews() {
        progress.center = convert(self.center, to: self.superview)
    }
    
    func setUrl(_ url: String) {
        setUrl(url, cache: true)
    }
    
    func setUrl(_ url: String, cache: Bool) {
        self.image = nil
        queue.cancelAllOperations()
        progress.startAnimating()
        queue.addOperation( { self.downloadImg(url, cache: true) } )
    }
    
    func downloadImg(_ url: String, cache: Bool) {
        var data : Data!
        if !cache {
            data = try! Data(contentsOf: URL(string: url)!)
        } else {
            var path = url
            path = path.replacingOccurrences(of: "/", with: "_")
            path = path.replacingOccurrences(of: "\\", with: "_")
            path = path.replacingOccurrences(of: ":", with: "_")
            path = NSHomeDirectory() + "/Documents/" + path
            
            if FileManager.default.fileExists(atPath: path) {
                data = try? Data(contentsOf: URL(fileURLWithPath: path))
            } else {
                data = try! Data(contentsOf: URL(string: url)!)
                try! data.write(to: URL(fileURLWithPath: path))
            }
        }
        
        mainQueue.addOperation( { self.showImg(data) } )
    }
    
    func showImg(_ data: Data) {
        if data.count == 0 {
            return;
        }
        
        self.image = UIImage(data: data)
        
        progress.stopAnimating()
    }
}
